hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -940,46 +954,53 @@ public String startCommit(String actionType, HoodieTableMetaClient metaClient) {
    */
   public void startCommitWithTime(String instantTime) {
     HoodieTableMetaClient metaClient = createMetaClient(true);
-    startCommitWithTime(instantTime, metaClient.getCommitActionType(), metaClient);
+    startCommitWithTime(Option.of(instantTime), metaClient.getCommitActionType(), metaClient);
   }
 
   /**
    * Completes a new commit time for a write operation (insert/update/delete/insert_overwrite/insert_overwrite_table) with specified action.
    */
   public void startCommitWithTime(String instantTime, String actionType) {
     HoodieTableMetaClient metaClient = createMetaClient(true);
-    startCommitWithTime(instantTime, actionType, metaClient);
+    startCommitWithTime(Option.of(instantTime), actionType, metaClient);
   }
 
   /**
    * Starts a new commit time for a write operation (insert/update/delete) with specified action.
    */
-  private void startCommitWithTime(String instantTime, String actionType, HoodieTableMetaClient metaClient) {
+  private String startCommitWithTime(Option<String> providedInstantTime, String actionType, HoodieTableMetaClient metaClient) {
     if (needsUpgrade(metaClient)) {
       // unclear what instant to use, since upgrade does have a given instant.
       executeUsingTxnManager(Option.empty(), () -> tryUpgrade(metaClient, Option.empty()));
     }
     CleanerUtils.rollbackFailedWrites(config.getFailedWritesCleanPolicy(),
         HoodieTimeline.COMMIT_ACTION, () -> tableServiceClient.rollbackFailedWrites(metaClient));
 
-    LOG.info("Generate a new instant time: {} action: {}", instantTime, actionType);
-    // check there are no inflight restore before starting a new commit.
-    HoodieTimeline inflightRestoreTimeline = metaClient.getActiveTimeline().getRestoreTimeline().filterInflightsAndRequested();
-    ValidationUtils.checkArgument(inflightRestoreTimeline.countInstants() == 0,
-        () -> "Found pending restore in active timeline. Please complete the restore fully before proceeding. As of now, "
-            + "table could be in an inconsistent state. Pending restores: "
-            + Arrays.toString(inflightRestoreTimeline.getInstantsAsStream().map(HoodieInstant::requestedTime).toArray()));
-
-    if (config.getFailedWritesCleanPolicy().isLazy()) {
-      this.heartbeatClient.start(instantTime);
-    }
+    txnManager.beginTransaction(Option.empty(), Option.empty());
+    String instantTime;
+    try {
+      instantTime = providedInstantTime.orElseGet(() -> createNewInstantTime(false));
+      LOG.info("Generate a new instant time: {} action: {}", instantTime, actionType);
+      // check there are no inflight restore before starting a new commit.
+      HoodieTimeline inflightRestoreTimeline = metaClient.reloadActiveTimeline().getRestoreTimeline().filterInflightsAndRequested();
```

Comment:
This timeline reloading is part of the locking now.  Should the transaction manager only guard `instantTime = providedInstantTime.orElseGet(() -> createNewInstantTime(false));`?

Reply:
If we are going to fail the writer on this condition then we should make sure the view is up to date within the transaction. Otherwise there is no guarantee that this writer's view of the timeline is up to date.

Reply:
The timeline reload is not event there before, does it mean it does not need to require the latest view of the timeline then or just a mistake?

Reply:
I think this is a mistake but I am not sure. My reason for thinking it may be a mistake is that there can be another writer that starts a restore between generating the metaclient and its first timeline reload so you may miss the restore. The question is whether this miss will cause any correctness issues or if this will just waste some work, this is the part I am unsure about. If it is just wasted work then we don't need the reload since that just forces more listing in the general case of normal operations.

Reply:
I think this could cause correctness issues but the comments in the restore code mention that you should stop all ingestion and table services before starting a restore so it seems like this path will not be hit if the user is following proper procedure.

Reply:
yeah, let's keep it as it is.